File: android/crime/src/main/java/com/dreampany/crime/ui/fragment/MoreFragment.java
package com.dreampany.crime.ui.fragment;

import androidx.databinding.ObservableArrayList;
import android.os.Bundle;
import androidx.recyclerview.widget.RecyclerView;
import android.view.Menu;
import android.view.View;

import com.dreampany.framework.databinding.FragmentItemsBinding;
import com.dreampany.framework.misc.ActivityScope;
import com.dreampany.framework.ui.fragment.BaseMenuFragment;
import com.dreampany.framework.util.SettingsUtil;
import com.dreampany.framework.util.ViewUtil;
import com.dreampany.crime.R;
import com.dreampany.crime.ui.adapter.MoreAdapter;
import com.dreampany.crime.ui.model.MoreItem;

import javax.inject.Inject;

import eu.davidea.flexibleadapter.FlexibleAdapter;
import eu.davidea.flexibleadapter.common.FlexibleItemDecoration;
import eu.davidea.flexibleadapter.common.SmoothScrollLinearLayoutManager;

/**
 * Created by Hawladar Roman on 5/24/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
@ActivityScope
public class MoreFragment extends BaseMenuFragment implements
        FlexibleAdapter.OnItemClickListener {

    private FragmentItemsBinding binding;
    private MoreAdapter adapter;

    @Inject
    public MoreFragment() {

    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_items;
    }

    @Override
    protected void onMenuCreated(Menu menu) {

    }

    @Override
    protected void onStartUi(Bundle state) {
        setTitle(R.string.more);
        binding = (FragmentItemsBinding) super.binding;
        initRecycler();
        binding.getRoot().post(() -> produceItems());
    }

    @Override
    protected void onStopUi() {

    }

    @Override
    public boolean onItemClick(View view, int position) {
        if (position != RecyclerView.NO_POSITION) {
            MoreItem item = adapter.getItem(position);
            showItem(item);
            return true;
        }
        return false;
    }

    private void initRecycler() {
        binding.setItems(new ObservableArrayList<>());
        adapter = new MoreAdapter(this);
        ViewUtil.setRecycler(
                binding.recycler,
                adapter,
                new SmoothScrollLinearLayoutManager(getContext()),
                null,
                new FlexibleItemDecoration(getContext())
                        .addItemViewType(R.layout.item_more, 0, 0, 0, 1)
                        //.withBottomEdge(false)
                        .withEdge(true)
        );
    }

    private void produceItems() {
        adapter.load();
    }

    private void showItem(MoreItem item) {
        switch (item.getType()) {
            case APPS:
                moreApps();
                break;
            case RATE_US:
                rateUs();
                break;
            case FEEDBACK:
                sendFeedback();
                break;
            default:
/*                UiTask<?, MoreType, ?, ?> task = new UiTask<>();
                task.setType(item.getType());
                AndroidUtil.openActivity(getParent(), ToolsActivity.class, task);*/
                break;
        }
    }

    private void moreApps() {
        SettingsUtil.moreApps(getActivity());
    }

    private void rateUs() {
        SettingsUtil.rateUs(getActivity());
    }

    private void sendFeedback() {
        SettingsUtil.feedback(getActivity());
    }
}

File: android/tools/src/main/java/com/dreampany/tools/ui/more/fragment/MoreFragment.kt
package com.dreampany.tools.ui.more.fragment

import android.os.Bundle
import com.dreampany.framework.data.model.Response
import com.dreampany.framework.inject.annote.ActivityScope
import com.dreampany.framework.misc.exts.disable
import com.dreampany.framework.misc.exts.moreApps
import com.dreampany.framework.misc.exts.rateUs
import com.dreampany.framework.ui.fragment.InjectFragment
import com.dreampany.tools.R
import com.dreampany.tools.data.enums.Action
import com.dreampany.tools.data.enums.State
import com.dreampany.tools.data.enums.Subtype
import com.dreampany.tools.data.enums.Type
import com.dreampany.tools.data.model.more.More
import com.dreampany.tools.databinding.RecyclerFragmentBinding
import com.dreampany.tools.ui.more.adapter.FastMoreAdapter
import com.dreampany.tools.ui.more.model.MoreItem
import com.dreampany.tools.ui.more.vm.MoreViewModel
import com.mikepenz.aboutlibraries.LibsBuilder
import kotlinx.android.synthetic.main.content_recycler_ad.view.*
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by roman on 20/3/20
 * Copyright (c) 2020 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@ActivityScope
class MoreFragment
@Inject constructor() : InjectFragment() {

    private lateinit var bind: RecyclerFragmentBinding
    private lateinit var adapter: FastMoreAdapter
    private lateinit var vm: MoreViewModel

    override val layoutRes: Int = R.layout.recycler_fragment

    override fun onStartUi(state: Bundle?) {
        initUi()
        initRecycler(state)
        if (adapter.isEmpty)
            vm.loadMores()
    }

    override fun onStopUi() {
    }

    override fun onSaveInstanceState(outState: Bundle) {
        if (::adapter.isInitialized) {
            var outState = outState
            outState = adapter.saveInstanceState(outState)
            super.onSaveInstanceState(outState)
            return
        }
        super.onSaveInstanceState(outState)
    }

    private fun initUi() {
        if (::bind.isInitialized) return
        bind = binding()
        vm = createVm(MoreViewModel::class)

        vm.subscribes(this, { this.processResponse(it) })

        bind.swipe.disable()
    }

    private fun initRecycler(state: Bundle?) {
        if (::adapter.isInitialized) return
        adapter = FastMoreAdapter(
            clickListener = { item: MoreItem ->
                Timber.v("MoreItem: %s", item.input.toString())
                onPressed(item.input)
            })
        adapter.initRecycler(state, bind.layoutRecycler.recycler)
    }

    private fun processResponse(response: Response<Type, Subtype, State, Action, List<MoreItem>>) {
        if (response is Response.Progress) {
            if (response.progress) showProgress() else hideProgress()
        } else if (response is Response.Error) {
            processError(response.error)
        } else if (response is Response.Result<Type, Subtype, State, Action, List<MoreItem>>) {
            Timber.v("Result [%s]", response.result)
            processResults(response.result)
        }
    }

    private fun processError(error: Throwable) {
        showDialogue(
            R.string.title_dialog_more,
            message = error.message,
            onPositiveClick = {

            },
            onNegativeClick = {

            }
        )
    }

    private fun processResults(result: List<MoreItem>?) {
        if (result != null) {
            adapter.addItems(result)
        }
    }

    private fun onPressed(item: More) {
        when (item.subtype) {
            Subtype.APPS -> {
                activity.moreApps(getString(R.string.id_google_play))
            }
            Subtype.RATE_US -> {
                activity.rateUs()
            }
            Subtype.FEEDBACK -> {
                activity.rateUs()
                //activity.moreApps(getString(R.string.id_google_play))
            }
            Subtype.LICENSE -> {
                activity.moreApps(getString(R.string.id_google_play))
            }
            Subtype.ABOUT -> {
                LibsBuilder().start(requireContext())
            }
        }

    }
}